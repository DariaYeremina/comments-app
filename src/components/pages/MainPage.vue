<template>
    <div class="wrapper">
        <VHeading>{{ $t('appTitle') }}</VHeading>
        <VSelect :options="getUsers"
                data-test="v-select"
                :placeholder="$t('select.placeholder')"
                @select="fetchPosts()"></VSelect>
        <VManageUserData v-if="getActiveUserId !== null"
                        data-test="v-manage-user-data"
                        @reload="fetchUsers"></VManageUserData>
        <VPostsList></VPostsList>
    </div>
</template>

<script>
import VHeading from '@/components/atoms/VHeading.vue';
import VSelect from '@/components/molecules/VSelect.vue';
import VPostsList from '@/components/organisms/VPostsList.vue';
import VManageUserData from '@/components/organisms/VManageUserData.vue';
import { mapGetters } from 'vuex';

export default {
  name: 'MainPage',
  components: {
    VHeading, VSelect, VPostsList, VManageUserData,
  },
  computed: {
    ...mapGetters('users', [
      'getUsers',
      'getActiveUserId',
    ]),
  },
  methods: {
    fetchPosts() {
      this.$store.dispatch('posts/getPostsPerUser');
    },
    fetchUsers() {
      this.$store.dispatch('users/getUsers');
    },
  },
  beforeRouteEnter(from, to, next) {
    next((vm) => {
      vm.fetchUsers();
    });
  },
};
</script>

<style lang="scss" scoped>
    .wrapper {
        width: 70vw;
        display: flex;
        flex-direction: column;
        align-items: center;
    }
</style>
