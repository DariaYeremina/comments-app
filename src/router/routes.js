import MainPage from '@/components/pages/MainPage.vue';

export default [
  {
    path: '/',
    name: 'Home',
    component: MainPage,
  },
];
