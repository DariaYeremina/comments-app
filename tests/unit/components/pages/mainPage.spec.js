import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import flushPromises from 'flush-promises';
import routes from '@/router/routes';
import MainPage from '@/components/pages/MainPage.vue';

const localVue = createLocalVue();
localVue.use(VueRouter);
localVue.use(Vuex);

const router = new VueRouter({ routes });

const users = {
  namespaced: true,
  getters: {
    getUsers: jest.fn().mockReturnValue([]),
    getActiveUserId: jest.fn().mockReturnValue(1),
  },
  actions: {
    getUsers: jest.fn(),
  },
};

const posts = {
  namespaced: true,
  actions: {
    getPostsPerUser: jest.fn(),
  },
};

const store = new Vuex.Store({
  modules: {
    users,
    posts,
  },
});

describe('MainPage component', () => {
  let wrapper;
  const app = document.createElement('div');
  app.setAttribute('data-app', true);
  document.body.append(app);

  beforeEach(() => {
    wrapper = mount(MainPage, {
      localVue,
      store,
      router,
      mocks: {
        $t: (msg) => msg,
      },
      sync: false,
      stubs: {
        VHeading: true,
        VSelect: true,
        VManageUserData: true,
        VPostsList: true,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Component renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Fetch users on BeforeRouteEnter', async () => {
    const from = {};
    const to = {};
    const next = jest.fn();
    const fetchUsersSpyOn = jest.spyOn(users.actions, 'getUsers');
    wrapper.vm.fetchUsers = jest.fn();

    wrapper.vm.$options.beforeRouteEnter[0](to, from, (cb) => cb(wrapper.vm));
    await flushPromises();

    expect(await wrapper.vm.fetchUsers).toHaveBeenCalledTimes(1);
  });

  it('Fetch posts on select event emitted by VSelect', async () => {
    const vSelect = wrapper.find('[data-test="v-select"]');
    const fetchPostsSpyOn = jest.spyOn(posts.actions, 'getPostsPerUser');

    vSelect.vm.$emit('select');
    await flushPromises();

    expect(await fetchPostsSpyOn).toHaveBeenCalledTimes(1);
  });

  it('Fetch users on reload event emitted by VManageUserData', async () => {
    const vManageUserData = wrapper.find('[data-test="v-manage-user-data"]');
    const fetchUsersSpyOn = jest.spyOn(users.actions, 'getUsers');

    vManageUserData.vm.$emit('reload');
    await flushPromises();

    expect(await fetchUsersSpyOn).toHaveBeenCalledTimes(1);
  });
});
