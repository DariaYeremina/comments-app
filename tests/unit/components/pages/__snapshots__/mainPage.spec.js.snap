// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage component Component renders correctly 1`] = `
<div
  class="wrapper"
>
  <vheading-stub>
    appTitle
  </vheading-stub>
   
  <vselect-stub
    data-test="v-select"
    itemid="id"
    itemtext="fullname"
    options=""
    placeholder="select.placeholder"
  />
   
  <vmanageuserdata-stub
    data-test="v-manage-user-data"
  />
   
  <vpostslist-stub />
</div>
`;
